icu4c/source/i18n/collation.cpp

```diff
@@ -20,15 +20,6 @@
 
 U_NAMESPACE_BEGIN
 
-// Some compilers don't care if constants are defined in the .cpp file.
-// MS Visual C++ does not like it, but gcc requires it. clang does not care.
-#ifndef _MSC_VER
-const uint8_t Collation::LEVEL_SEPARATOR_BYTE;
-const uint8_t Collation::MERGE_SEPARATOR_BYTE;
-const uint32_t Collation::ONLY_TERTIARY_MASK;
-const uint32_t Collation::CASE_AND_TERTIARY_MASK;
-#endif
```

Comment:
The comment says that gcc requires these definitions. Shouldn't we keep these then? On the other hand, the build bots seem happy.

Reply:
I don't understand that comment since these are defined in a header file unconditionally. https://github.com/unicode-org/icu/blob/3cefbd55c7aafbb29bc67aa9a27df9bf13293a5a/icu4c/source/i18n/collation.h#L32

Reply:
C++ lets you declare class constants, and assign values, in the class declaration, often in a .h file. That goes together with a *definition*, normally in a .cpp file. It *at least used to be* that gcc required these while MSVC failed with these definitions. I see that our current build bots are green, but I am nervous about older MSVC versions. @jefgen ?

Reply:
If a value is assigned to `TERMINATOR_BYTE` in a header file, what is left to define in the implementation file?

Reply:
> If a value is assigned to `TERMINATOR_BYTE` in a header file, what is left to define in the implementation file?

The storage location. The value should mostly be inlined, but a compiler may generate code that refers to its storage address.

Reply:
> I see that our current build bots are green, but I am nervous about older MSVC versions.

Well, FWIW, in terms of older versions of the MSVC compiler I think that the version that ships with VS 2015 (19.0) would be the oldest that would work for ICU now that it requires and uses C++11.

For example, support for the new C++11 type `char16_t` was [added](https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=msvc-170#c-compiler-features) in [VS 2015 (19.0)](https://en.cppreference.com/w/cpp/compiler_support/11) -- so that would be the minimum needed to compile and test ICU now.

It might be possible to hack/modify the code to compile with an older version of MSVC, but I think that would likely be "not supported" (at least for newer versions of the ICU library).

We do have a [build bot ](https://github.com/unicode-org/icu/blob/main/.ci-builds/.azure-pipelines.yml#L467)that uses the VS 2015 compiler toolchain, so if the builds are green, then this works on VS 2015.

Reply:
Thanks Jeff... Actually, I asked the wrong question :-(
Some version of MSVC failed with these definitions, but removing them altogether removes them for all compilers, even gcc which used to require them. So the real question is about old versions of *gcc*.

Maybe we are sufficiently covered. If this shows up again, then we should probably go full C++11 and change the declarations to using `constexpr` not merely `const`.

Reply:
Marking this as resolved, since it sounds like its alright to remove this moving forward.